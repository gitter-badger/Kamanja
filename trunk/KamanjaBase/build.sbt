name := "KamanjaBase"

version := "1.0"

scalaVersion := "2.10.4"

libraryDependencies += "log4j" % "log4j" % "1.2.17"

libraryDependencies += "org.json4s" %% "json4s-native" % "3.2.9" 

libraryDependencies += "org.json4s" %% "json4s-jackson" % "3.2.9" 

libraryDependencies += "com.google.guava" % "guava" % "16.0.1" 

libraryDependencies += "com.google.code.findbugs" % "jsr305" % "1.3.9"
