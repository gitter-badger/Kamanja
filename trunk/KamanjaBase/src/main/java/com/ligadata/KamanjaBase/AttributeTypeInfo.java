package com.ligadata.KamanjaBase;

public class AttributeTypeInfo {

	String name;
	int Index;
	TypeCategory typeCategory; // from TypeCategory
	byte valTypeId;
	byte keyTypeId;
	long valSchemaId;

	public AttributeTypeInfo(String name, int Index, TypeCategory typeCategory,
			byte valTypeId, byte keyTypeId, long valSchemaId) {
		this.name = name;
		this.Index = Index;
		this.typeCategory = typeCategory;
		this.valTypeId = valTypeId;
		this.valSchemaId = valSchemaId;
		this.keyTypeId = keyTypeId;
	}

	public enum TypeCategory {
		INT(0), STRING(1), FLOAT(2), DOUBLE(3), LONG(4), BYTE(5), CHAR(6), CONTAINER(
				1001), MAP(1002), ARRAY(1003);
		private int value;

		private TypeCategory(int value) {
			this.value = value;
		}

		public int getValue(){
			return this.value;
		}
	}

	public boolean IsContainer() {
		return (typeCategory == TypeCategory.CONTAINER);
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public int getIndex() {
		return Index;
	}

	public void setIndex(int index) {
		Index = index;
	}

	public TypeCategory getTypeCategary() {
		return (TypeCategory) typeCategory;
	}

	public void setTypeCategary(TypeCategory typeCategory) {
		this.typeCategory = typeCategory;
	}

	public byte getValTypeId() {
		return valTypeId;
	}

	public void setValTypeId(byte valTypeId) {
		this.valTypeId = valTypeId;
	}

	public byte getKeyTypeId() {
		return keyTypeId;
	}

	public void setKeyTypeId(byte keyTypeId) {
		this.keyTypeId = keyTypeId;
	}

	public long getValSchemaId() {
		return valSchemaId;
	}

	public void setValSchemaId(long valSchemaId) {
		this.valSchemaId = valSchemaId;
	}

}
