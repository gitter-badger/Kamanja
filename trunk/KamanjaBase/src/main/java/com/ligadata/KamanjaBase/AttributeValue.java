package com.ligadata.KamanjaBase;

public class AttributeValue {

	String valueType = "";
	Object value = "";

	public String getValueType() {
		return this.valueType;
	}

	public void setValueType(String valType) {
		this.valueType = valType;
	}

	public Object getValue() {
		return this.value;
	}

	public void setValue(Object val) {
		this.value = val;
	}
}
