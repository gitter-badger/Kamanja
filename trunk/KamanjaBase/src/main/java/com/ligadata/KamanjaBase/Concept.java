package com.ligadata.KamanjaBase;

//FIXME:: Decide whether we need to generate concept class for every concept or use generic class like this
public class Concept implements ContainerOrConcept {
	String key = null;
	String valueType = null;
	Object value = null;

	Concept(String key, String valueType, Object value) {
		this.key = key;
		this.valueType = valueType;
		this.value = value;
	}
}