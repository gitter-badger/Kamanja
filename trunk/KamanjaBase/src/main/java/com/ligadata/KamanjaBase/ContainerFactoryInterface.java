package com.ligadata.KamanjaBase;

public abstract class ContainerFactoryInterface implements
		ContainerOrConceptFactory {

	enum ContainerType {
		MESSAGE, CONTAINER;
	}

	final public String getFullTypeName() {
		return (getTypeNameSpace() + "." + getTypeName());
	}

	final public boolean hasPrimaryKey() {
		String[] pKeys = getPrimaryKeyNames();
		return (pKeys != null && pKeys.length > 0);
	}

	final public boolean hasPartitionKey() {
		String[] pKeys = getPartitionKeyNames();
		return (pKeys != null && pKeys.length > 0);
	}

	final public boolean hasTimePartitionInfo() {
		TimePartitionInfo tmInfo = getTimePartitionInfo();
		return (tmInfo != null && tmInfo.timePartitionType != TimePartitionInfo.TimePartitionType.NONE);
	}

	public abstract String getTypeNameSpace();

	public abstract String getTypeName();

	public abstract String getTypeVersion();

	public abstract TimePartitionInfo getTimePartitionInfo();

	public abstract ContainerType getContainerType(); // ContainerType is enum of MESSAGE or CONTAINER

	public abstract boolean isFixed();

	public abstract String getSchema();

	public abstract String[] getPrimaryKeyNames();

	public abstract String[] getPartitionKeyNames();

	public abstract ContainerInterface createInstance();

}
