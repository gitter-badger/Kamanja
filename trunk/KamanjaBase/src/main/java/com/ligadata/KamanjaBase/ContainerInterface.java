package com.ligadata.KamanjaBase;

public abstract class ContainerInterface extends ContainerOrConcept {
	long transactionId = 0;
	long timePartitionData = 0;
	int rowNumber = 0;

	public ContainerInterface(ContainerFactoryInterface factory) {
		super(factory);
	}

	final ContainerFactoryInterface getContainerFactory() {
		return ((ContainerFactoryInterface) getFactory());
	}

	final public boolean hasPrimaryKey() {
		return getContainerFactory().hasPrimaryKey();
	}

	final public boolean hasPartitionKey() {
		return getContainerFactory().hasPartitionKey();
	}

	final public boolean hasTimePartitionInfo() {
		return getContainerFactory().hasTimePartitionInfo();
	}

	final public long getTransactionId() {
		return transactionId;
	}

	final public long getTimePartitionData() {
		return timePartitionData;
	}

	final public int getRowNumber() {
		return rowNumber;
	}

	final public int getSchemaId() {
		return getContainerFactory().getSchemaId();
	}

	final public ContainerFactoryInterface.ContainerType getContainerType() {
		return getContainerFactory().getContainerType();
	}

	final public boolean isFixed() {
		return getContainerFactory().isFixed();
	}

	final public String getAvroSchema() {
		return getContainerFactory().getAvroSchema();
	}

	final public String[] getPrimaryKeyNames() {
		return getContainerFactory().getPrimaryKeyNames();
	}

	final public String[] getPartitionKeyNames() {
		return getContainerFactory().getPartitionKeyNames();
	}

	final public TimePartitionInfo getTimePartitionInfo() {
		return getContainerFactory().getTimePartitionInfo();
	}

	final public void setTransactionId(long transId) {
		this.transactionId = transId;
	}

	final public void setTimePartitionData(long timePrtData) {
		this.timePartitionData = timePrtData;
	}

	final public void setRowNumber(int rowNbr) {
		this.rowNumber = rowNbr;
	}

	public abstract String[] getPrimaryKey();

	public abstract String[] getPartitionKey();

	public abstract void save();

	public abstract ContainerOrConcept Clone();

	public abstract String[] getAttributeNames();

	public abstract AttributeTypeInfo[] getAttributeTypes();

	public abstract Object get(String key); // Return (value, type)

	public abstract Object get(int index); // Return (value, type)

	public abstract Object getOrElse(String key, Object defaultVal); // Return
																		// (value,
																		// type)

	public abstract Object getOrElse(int index, Object defaultVal); // Return
																	// (value,
																	// type)

	public abstract AttributeValue[] getAllAttributeValues(); // Has (name,
																// value, type))

	public abstract void set(String key, Object value);

	public abstract void set(int index, Object value);

	public abstract void set(String key, Object value, String valTyp);

	public abstract java.util.Iterator<AttributeValue> getAttributeNameAndValueIterator();
}
