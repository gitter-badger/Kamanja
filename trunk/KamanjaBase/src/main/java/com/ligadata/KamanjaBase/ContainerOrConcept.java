package com.ligadata.KamanjaBase;

//FIXME:: Need to change ContainerOrConcept name to some meaningful name

public interface ContainerOrConcept {

}
