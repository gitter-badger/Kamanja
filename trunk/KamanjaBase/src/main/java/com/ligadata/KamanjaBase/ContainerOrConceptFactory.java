package com.ligadata.KamanjaBase;

public interface ContainerOrConceptFactory {

}
