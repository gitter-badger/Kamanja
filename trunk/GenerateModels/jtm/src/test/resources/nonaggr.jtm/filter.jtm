{
  "eplquery": "select * from com.ligadata.kamanja.samples.messages where score != -1 and score < 100",

  "packagename": "com.ligadata.jtm.test.filter",
  "version": "0.0.1",
  "inputs": [
    { "typename": "com.ligadata.kamanja.samples.messages.msg1" }
  ],

  "filters": [
    {"expression": "score != -1 && score < 100"}
  ],

  "outputs": [
    { "typename":"com.ligadata.kamanja.samples.messages.msg1" }
  ]
}
