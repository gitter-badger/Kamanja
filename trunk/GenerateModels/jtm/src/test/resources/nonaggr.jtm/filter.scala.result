/*
 * Copyright 2016 ligaDATA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ligadata.jtm.test.filter

import com.ligadata.KamanjaBase._
import com.ligadata.KvBase.TimeRange
import com.ligadata.kamanja.metadata.ModelDef;

// Normalize the name
import com.ligadata.kamanja.samples.messages.{msg1 ⇒ input1}
import com.ligadata.kamanja.samples.messages.{msg1 ⇒ output1}

class Factory(modelDef: ModelDef, nodeContext: NodeContext) extends ModelInstanceFactory(modelDef, nodeContext) {
  // if more than 1 input we have to find the correct instance
  override def isValidMessage(msg: MessageContainerBase): Boolean = return msg.isInstanceOf[msg1]
  override def createModelInstance(): ModelInstance = return new Model(this)

  // Provided in json
  override def getModelName: String = "com.ligadata.jtm.test.filter"
  override def getVersion: String = "0.0.1"
  
  override def createResultObject(): ModelResultBase = new MappedModelResults()
}

class Model(factory: ModelInstanceFactory) extends ModelInstance(factory) {

   override def execute(txnCtxt: TransactionContext, outputDefault: Boolean): ModelResultBase = {

     // Get input, if more than 1 input we have to find the correct instance
     var input1 : msg1 =  txnCtxt.getMessage().asInstanceOf[input1]

     // Filter condition, shortcut evaluation
     if(!(score != -1 && score < 100))
        return null;

     /// Produce output
     var result: Array[Result] =
        Array[Result](new Result("Id", output1.id), new Result("Name", output1.Name), new Result("Score", output1.score))

     // Return output
     return factory.createResultObject().asInstanceOf[MappedModelResults].withResults(result)
   }
}