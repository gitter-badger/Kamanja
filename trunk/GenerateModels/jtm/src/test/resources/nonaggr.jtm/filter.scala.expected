
/*
 * Copyright 2016 ligaDATA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ligadata.jtm.test.filter

import com.ligadata.KamanjaBase._
import com.ligadata.KvBase.TimeRange
import com.ligadata.kamanja.metadata.ModelDef;

import com.ligadata.kamanja.samples.messages.{msg1 ⇒ input0}
import com.ligadata.kamanja.samples.messages.{msg1 ⇒ output0}

class Factory(modelDef: ModelDef, nodeContext: NodeContext) extends ModelInstanceFactory(modelDef, nodeContext) {
  // if more than 1 input we have to find the correct instance
  override def isValidMessage(msg: MessageContainerBase): Boolean = return msg.isInstanceOf[input0]
  override def createModelInstance(): ModelInstance = return new Model(this)

  // Provided in json
  override def getModelName: String = "filter"
  override def getVersion: String = "0.0.1"

  override def createResultObject(): ModelResultBase = new MappedModelResults()
}


class Model(factory: ModelInstanceFactory) extends ModelInstance(factory) {

   override def execute(txnCtxt: TransactionContext, outputDefault: Boolean): ModelResultBase = {

    var input0: input0 = txnCtxt.getMessage().asInstanceOf[input0]


    {
      using input0
      if(!(score != -1 && score < 100))
        return null;
    }
    var result: Array[Result] = Array[Result](new Result("id", input0.id), new Result("name", input0.name), new Result("score", input0.score), new Result("transactionid", input0.transactionid))
    factory.createResultObject().asInstanceOf[MappedModelResults].withResults(result)


   }


}
